com/williamfiset/datastructures/binarysearchtree/SplayTree.java

```diff
@@ -0,0 +1,274 @@
+import com.williamfiset.datastructures.utils.TreePrinter;
+
+import java.util.ArrayList;
+import java.util.Scanner;
+
+
+class BinaryTree<T extends Comparable<T>> implements TreePrinter.PrintableNode {
+    private BinaryTree<T> left, right;
+    private T data;
+
+    public BinaryTree(T data) {
+        this.data = data;
+    }
+
+    @Override
+    public BinaryTree<T> getLeft() {
+        return left;
+    }
+
+    @Override
+    public BinaryTree<T> getRight() {
+        return right;
+    }
+
+    @Override
+    public String getText() {
+        return data.toString();
+    }
+    public void setRight(BinaryTree<T> right) {
+        this.right = right;
+        //   right.parent = this;
+    }
+    public void setLeft(BinaryTree<T> left) {
+        this.left = left;
+        // left.parent = this;
+    }
+    public T getData() {
+        return data;
+    }
+
+    public void setData(T data) {
+        this.data = data;
+    }
+
+    @Override
+    public String toString() {
+
+        return TreePrinter.getTreeDisplay(this);
+    }
+
+
+}
+
+/**Note : Each Operation of a SplayTree tries returns the new splayed root after the operation **/
+public class SplayTree<T extends Comparable<T>> {
+
+    private BinaryTree<T> root;
+
+    public SplayTree() {
+        this.root = null;
+    }
+
+    public SplayTree(BinaryTree<T> root) {
+        this.root = root;
+    }
+
+    public BinaryTree<T> getRoot() {
+        return root;
+    }
+
+    private BinaryTree<T> rightRotate(BinaryTree<T> x) {
+        BinaryTree<T> p = x.getLeft();
+        x.setLeft(p.getRight());
+        p.setRight(x);
+        return p;
+    }
+
+    private BinaryTree<T> leftRotate(BinaryTree<T> x) {
+        BinaryTree<T> p = x.getRight();
+        x.setRight(p.getLeft());
+        p.setLeft(x);
+        return p;
+    }
+
+    private BinaryTree<T> splayUtil(BinaryTree<T> root, T key) {
```

Comment:
I think this can be renamed just "splay" since it doesn't overlap with the other splay method function definition.

Reply:
Can be.  I can tell you why splayUtil() made sense to me. Then u decide.
1. The other splay Function actually uses this function.
2. Calling splayUtil() alone wont be required ever without calling splay().
so i thought its basically just a utility function of the original splay(node) operation.Therefore the name splayUtil.

Reply:
sg